src/LinAlg/hiopLinSolverSparseGinkgo.cpp

```diff
@@ -0,0 +1,376 @@
+    }
+}
+
+
+std::shared_ptr<gko::LinOpFactory> setup_solver_factory(std::shared_ptr<const gko::Executor> exec,
+        std::shared_ptr<gko::matrix::Csr<double, int>> mtx) {
+
+    auto preprocessing_fact = gko::share(gko::reorder::Mc64<double, int>::build().on(exec));
+    auto preprocessing = gko::share(preprocessing_fact->generate(mtx));
+
+    auto lu_fact = gko::share(gko::factorization::Glu<double, int>::build_reusable()
+            .on(exec, mtx.get(), preprocessing.get()));
+    auto inner_solver_fact = gko::share(gko::preconditioner::Ilu<>::build()
+            .with_factorization_factory(lu_fact)
+            .on(exec));
+    auto solver_fact = gko::share(gko::solver::Gmres<>::build()
+            .with_criteria(
+                gko::stop::Iteration::build()
+                    .with_max_iters(200u)
+                    .on(exec),
+                gko::stop::ResidualNorm<>::build()
+                    .with_baseline(gko::stop::mode::absolute)
+                    .with_reduction_factor(1e-8)
+                    .on(exec))
+            .with_krylov_dim(10u)
+            .with_preconditioner(inner_solver_fact)
+            .on(exec));
+
+    auto reusable_factory = gko::share(gko::solver::ScaledReordered<>::build()
+            .with_solver(solver_fact)
+            .with_reordering(preprocessing)
+            .on(exec));
+    return reusable_factory;
+}
+
+
+}
+
+
+  hiopLinSolverIndefSparseGinkgo::hiopLinSolverIndefSparseGinkgo(const int& n, const int& nnz, hiopNlpFormulation* nlp)
+    : hiopLinSolverSymSparse(n, nnz, nlp),
+      index_covert_CSR2Triplet_{nullptr},index_covert_extra_Diag2CSR_{nullptr},
+      n_{n}, nnz_{0}
+  {}
+
+  hiopLinSolverIndefSparseGinkgo::~hiopLinSolverIndefSparseGinkgo()
+  {
+    if(index_covert_CSR2Triplet_)
+      delete [] index_covert_CSR2Triplet_;
+    if(index_covert_extra_Diag2CSR_)
+      delete [] index_covert_extra_Diag2CSR_;
+  }
+
+  void hiopLinSolverIndefSparseGinkgo::firstCall()
+  {
+    assert(n_==M_->n() && M_->n()==M_->m());
+    assert(n_>0);
+
+    exec = gko::ReferenceExecutor::create(); //gko::CudaExecutor::create(0, gko::OmpExecutor::create());
+
+    mtx = transferTripletToCSR(exec, n_, M_, &index_covert_CSR2Triplet_, &index_covert_extra_Diag2CSR_);
+    nnz_ = mtx->get_num_stored_elements();
+
+    reusable_factory = setup_solver_factory(exec, mtx);
+  }
+
+  int hiopLinSolverIndefSparseGinkgo::matrixChanged()
+  {
+    assert(n_==M_->n() && M_->n()==M_->m());
+    assert(n_>0);
+
+    nlp_->runStats.linsolv.tmFactTime.start();
+
+    if( !mtx ){
+      this->firstCall();
+    }else{
+      update_matrix(M_, mtx, index_covert_CSR2Triplet_, index_covert_extra_Diag2CSR_);
+    }
+
+    gko_solver = gko::share(reusable_factory->generate(mtx));
+    nlp_->runStats.linsolv.tmInertiaComp.start();
+    int negEigVal = nFakeNegEigs_;
+    nlp_->runStats.linsolv.tmInertiaComp.stop();
+    return negEigVal;
```

Comment:
Can we add a list and some short explanations about the possible return flags from ginkgo?
For example, if ginkgo failed to factorize the matrix due to some numerical issue, we have something like `gko_solver.info[0] = -1`.

Reply:
@cnpetra and @nychiang: Are there defined error codes for HiOp or for now we are using `-1` when a function fails?